gui/src/NodeSelector.tsx

```diff
@@ -0,0 +1,196 @@
+    return JSON.stringify({ id: id, type: type });
+};
+
+const NodeItem = ({ item = [] }: { item: any[] }) => {
+    const [id, label, items = [], nodeType, _] = item;
+    return (
+        <TreeItem
+            key={id}
+            nodeId={buildId(id, nodeType)}
+            label={
+                <Box sx={treeItemLabelSx}>
+                    <Datanode fontSize="small" color="primary" />
+                    {label}
+                </Box>
+            }
+        />
+    );
+};
+
+const PipelineItem = ({ item = [] }) => {
+    const [id, label, items = [], nodeType, _] = item;
+
+    return (
+        <TreeItem
+            key={id}
+            nodeId={buildId(id, nodeType)}
+            label={
+                <Box sx={treeItemLabelSx}>
+                    <Pipeline fontSize="small" color="primary" />
+                    {label}
+                </Box>
+            }
+            sx={ParentTreeItemSx}
+        >
+            {items &&
+                items.map((item: any) => {
+                    return <NodeItem item={item} />;
+                })}
+        </TreeItem>
+    );
+};
+
+const ScenarioItem = ({ item = [] }: { item: any[] }) => {
+    const [id, label, items = [], nodeType, _] = item;
+    return (
+        <TreeItem
+            key={id}
+            nodeId={buildId(id, nodeType)}
+            label={
+                <Box sx={treeItemLabelSx}>
+                    <ScenarioIcon fontSize="small" color="primary" />
+                    {label}
+                </Box>
+            }
+            sx={ParentTreeItemSx}
+        >
+            {items &&
+                items.map((item: any) => {
+                    const [id, label, items = [], nodeType, _] = item;
+                    return nodeType === NodeType.PIPELINE ? <PipelineItem item={item} /> : <NodeItem item={item} />;
+                })}
+        </TreeItem>
+    );
+};
+
+const CycleItem = ({ item }: { item: any[] }) => {
+    const [id, label, items = [], nodeType, _] = item;
+    return (
+        <TreeItem
+            key={id}
+            nodeId={buildId(id, nodeType)}
+            label={
+                <Box sx={treeItemLabelSx}>
+                    <Cycle fontSize="small" color="primary" />
+                    {label}
+                </Box>
+            }
+            sx={ParentTreeItemSx}
+        >
+            {items &&
+                items.map((item: any) => {
+                    const [id, label, items = [], nodeType, _] = item;
+                    return nodeType === NodeType.SCENARIO ? <ScenarioItem item={item} /> : <NodeItem item={item} />;
+                })}
+        </TreeItem>
+    );
+};
+
+const NodeSelector = (props: NodeSelectorProps) => {
+    const { id = "", datanodes } = props;
+
+    const dispatch = useDispatch();
+    const module = useModule();
+
+    const onSelect = useCallback((e: React.SyntheticEvent | undefined, nodeIds: string) => {
+        const selected = (nodeIds && JSON.parse(nodeIds)) || "";
```

Comment:
it might be simpler to use a data-type attribute on the treeItem that we can retrieve from e.currentTarget